### PR TITLE
Add vertex related functions in corefinement visitor

### PR DESCRIPTION
**Pull request by sloriot**
## Summary of Changes

Add function to track creation and copy of new vertices during corefinement based operations
## Release Management

* Affected package(s): PMP
* Issue(s) solved (if any):  #5835 
* Feature/Small Feature (if any): [Small_Features/Coref_more_visitor_functions_for_vertices](https://cgal.geometryfactory.com/CGAL/Members/wiki/Features/Small_Features/Coref_more_visitor_functions_for_vertices) -- pre-approved Andreas Fabri 11:03, 22 July 2021 (CEST) 
* Link to compiled documentation:
  *  [`intersection_point_detected()`](https://cgal.github.io/5861/v1/Polygon_mesh_processing/classPMPCorefinementVisitor.html#a2423b02894bde65bd4f0abdd518d8f70)
  * [`new_vertex_added()`](https://cgal.github.io/5861/v1/Polygon_mesh_processing/classPMPCorefinementVisitor.html#a55b5f3047aadd61f1b1ed3805685835e)
  * [`after_vertex_copy()`](https://cgal.github.io/5861/v1/Polygon_mesh_processing/classPMPCorefinementVisitor.html#aa2a5afadd444a68d968b3214b9ac09f5)
  * [`before_vertex_copy()`](https://cgal.github.io/5861/v1/Polygon_mesh_processing/classPMPCorefinementVisitor.html#a194d63ade8a168bdaf031694388d7b19)
* License and copyright ownership: GPL/GF

### CODE REVIEW

**Comment by afabri** [comment]
Why are parameters calles `tm_f`  and `tm_e`?   In the function `corefine()`  the meshes are calles `tm1` and `tm2`.

**Comment by sloriot** [comment]
> Why are parameters called `tm_f` and `tm_e`? In the function `corefine()` the meshes are calles `tm1` and `tm2`.
> For `new_vertex_added()` I am wondering if it would be helpful to also say if it is `tm1` or `tm2`.
> In `new_vertex_added()` you do not specify what `h_e` is.
> 

Citing the doc: "The intersection is detected using a face of `tm_f` and an edge of `tm_e`"


> On the `CorefinementVisitor` [concept page](https://cgal.github.io/5861/v1/Polygon_mesh_processing/classPMPCorefinementVisitor.html) we mix `TriangleMesh` and `Triangle_mesh` in the API. This is not related to this PR, but while you are at it maybe just unify.


Already fixed in 6d11998

**Review by afabri — APPROVED** [review]

File: Polygon_mesh_processing/doc/Polygon_mesh_processing/Concepts/PMPCorefinementVisitor.h [hunk 1]
```diff
@@ -40,14 +40,46 @@ typedef unspecified_type halfedge_descriptor;
...
+  void edge_split(halfedge_descriptor hnew, Triangle_mesh& tm);
   /// called when the split of the halfedge `h` passed at the later call to `before_edge_split()` is finished.
   void after_edge_split();
   ///  called when a new edge has been added to triangulate a face. The face triangulated is `f_split`
   ///  in the last call to `before_subface_creations(f_split, tm)`. There is only one call per edge.
-  void add_retriangulation_edge(halfedge_descriptor h, TriangleMesh& tm);
+  void add_retriangulation_edge(halfedge_descriptor h, Triangle_mesh& tm);
+/// @}
+
+/// @name Function used by corefine() when a new vertex is created
```
Comment (MaelRL) [inline Polygon_mesh_processing/doc/Polygon_mesh_processing/Concepts/PMPCorefinementVisitor.h:53]:
```suggestion
/// @name Functions used by corefine() when a new vertex is created.
```

File: Polygon_mesh_processing/doc/Polygon_mesh_processing/Concepts/PMPCorefinementVisitor.h [hunk 2]
```diff
@@ -40,14 +40,46 @@ typedef unspecified_type halfedge_descriptor;
...
+  /// \param i_id the id of the intersection point, starting at 0. Ids are consecutive.
+  /// \param sdim indicates the dimension of the simplex part of the face that is intersected by the edge
+  ///             (0 for a vertex, 1 for an edge, 2 for the interior of the face)
+  /// \param h_f a halfedge from `tm_f` indicating the simplex intersected:
+  ///            if `sdim==0` the target of `h_f` is the intersection point,
+  ///            if `sdim==1` the edge of `h_f` contains the intersection point in its interior,
+  ///            if `sdim==2` the face of `h_f` contains the intersection point in its interior.
+  /// \param h_e a halfedge from `tm_e`
+  /// \param is_target_coplanar `true` iif the target of `h_e` is the intersection point
+  /// \param is_source_coplanar `true` iif the source of `h_e` is the intersection point
```
Comment (MaelRL) [inline Polygon_mesh_processing/doc/Polygon_mesh_processing/Concepts/PMPCorefinementVisitor.h:67]:
```suggestion
  /// \param is_target_coplanar `true` iff the target of `h_e` is the intersection point
  /// \param is_source_coplanar `true` iff the source of `h_e` is the intersection point
```

File: Polygon_mesh_processing/doc/Polygon_mesh_processing/Concepts/PMPCorefinementVisitor.h [hunk 3]
```diff
@@ -40,14 +40,46 @@ typedef unspecified_type halfedge_descriptor;
   /// the split of that edge. If `edge_split(h_i, tm)` is called for `i=1` to `n`,
   /// `h_1`, `h_2`, ... ,`h_n`, `h` is the sequence of halfedges representing the
   /// edge split (with the same initial orientation). There is only one call per edge.
-  void before_edge_split(halfedge_descriptor h, TriangleMesh& tm);
+  void before_edge_split(halfedge_descriptor h, Triangle_mesh& tm);
   /// called when a new split is done. The target of `hnew` is a new split vertex. There is only one call per edge.
-  void edge_split(halfedge_descriptor hnew, TriangleMesh& tm);
+  void edge_split(halfedge_descriptor hnew, Triangle_mesh& tm);
```
Comment (MaelRL) [inline Polygon_mesh_processing/doc/Polygon_mesh_processing/Concepts/PMPCorefinementVisitor.h:45]: some functions take a `&` and others take a `const&`

File: Polygon_mesh_processing/doc/Polygon_mesh_processing/Concepts/PMPCorefinementVisitor.h [hunk 4]
```diff
@@ -60,23 +92,29 @@ typedef unspecified_type halfedge_descriptor;
...
-  void intersection_edge_copy(halfedge_descriptor h_src1, const TriangleMesh& tm_src1,
-                              halfedge_descriptor h_src2, const TriangleMesh& tm_src2,
-                              halfedge_descriptor h_tgt,  TriangleMesh& tm_tgt);
+  void intersection_edge_copy(halfedge_descriptor h_src1, const Triangle_mesh& tm_src1,
+                              halfedge_descriptor h_src2, const Triangle_mesh& tm_src2,
+                              halfedge_descriptor h_tgt,  Triangle_mesh& tm_tgt);
+  ///called before vertex `v_src` from `tm_src` is copied in `tm_tgt`
+  void before_vertex_copy(vertex_descriptor v_src, const Triangle_mesh& tm_src, Triangle_mesh& tm_tgt);
+  ///called after vertex `v_src` from `tm_src` is copied in `tm_tgt`. The new vertex is `v_tgt`.
+  /// The point has already been put in the vertex point map.
```
Comment (MaelRL) [inline Polygon_mesh_processing/doc/Polygon_mesh_processing/Concepts/PMPCorefinementVisitor.h:116]:
```suggestion
  /// called before vertex `v_src` from `tm_src` is copied in `tm_tgt`
  void before_vertex_copy(vertex_descriptor v_src, const Triangle_mesh& tm_src, Triangle_mesh& tm_tgt);
  /// called after vertex `v_src` from `tm_src` is copied in `tm_tgt`. The new vertex is `v_tgt`.
  /// The point has already been put in the vertex point map.
```

**Comment by sloriot** [comment]
Successfully tested in CGAL-5.4-Ic-28

**Pull request merged**